protocol/lib/abci/util.go

```diff
@@ -0,0 +1,18 @@
+package abci
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+// RunCached wraps a function with a cache context and writes the cache to the context if the
+// function call succeeds. If the function call fails, the cache is discarded.
+func RunCached(c sdk.Context, f func(sdk.Context) error) error {
```

Comment:
Should we also do the following, similar to Axelar [RunCached](https://github.com/axelarnetwork/axelar-core/blob/e9287fd765f7e501238ab85395ba874d5d89b52c/utils/abci.go#L19) function:
- Add recovery. Since  x/delay-msg `DispatchMessage` happens in the `EndBlocker`, a panic in `f(ctx)`, if not recovered leads to consensus failure. We should also log error in this case, similar to the behavior when `runTx` panic is recoverred

Reply:
I added panic recovery and check in the test cases that events are added to the context's event manager only when the function succeeeds. PTAL!